### DISCUSSION TITLE
Block "Buffer Hints"

### DISCUSSION THREAD

**Discussion by etschneider** [FutureSDR/FutureSDR#94] [Ideas]
Just some ideas for the future...

Should there be a mechanism for a block to provide extra information to the flowgraph about its buffer needs?

For example minimum required buffer sizes. Perhaps even the buffer type (e.g. Vulkan, etc). This would reduce the need to use `connect_stream_with_type()`. 

A use case:

I want to make a "trigger" block that captures some pre-trigger samples. To do this it has to maintain an amount of "history" and the input buffer is the natural place to keep that history. I can manually specify a big buffer with `connect_stream_with_type()`, but it seems inelegant.

**Comment by bastibl** [comment]
I was also thinking about this. We definitely need _something_. My current vague idea is to make the block smart and keep the runtime dumb. Already in `init()`, the block has access to its stream IO. So it could check some buffer parameters (e.g., history and maybe other things) and fail if things cannot work (e.g., FIR) or adapt (e.g., reduce the pre-triggered sample size).

Everything else gets probably much more complicated, since we'd need a constraint solver to construct the graph, if every block can make a wish and define acceptable ranges for buffer parameters :-) But maybe there are libraries available.

**Reply by etschneider** [reply]
Sorry about the latency in my response... The recent discussion about time triggered blocks reminded me of this thread.

I'm not sure I understand what you are suggesting. Is it to just make it so blocks can fail more gracefully in `init()` than in `work()`?  And leave it to the user (creator of the flowgraph) to ensure that the buffers are configured properly? Isn't that pretty much the way things are now?

I think it would be preferable for block writers that had special buffer needs to be able to provide hints to the runtime using the specifics of their implementations (and instantiation parameters). Having the users of the blocks need know what special buffer parameters are required doesn't seem very friendly.

Consider a scope block. It will likely want a relatively large buffer "history" for allow it to display a large number of samples. It doesn't seem right that a user of a scope block should be required to use `connect_stream_with_type()` anytime they want to connect a scope display to a block.

The size hint feature doesn't seem that difficult to me, but maybe I'm missing something. Just have each block specify a minimum buffer size and then the runtime uses the max() of all the values collected from all the connected blocks. This alone would be very useful, IMO.

I guess the main implementation problem is that the buffer creation would have to happen *after* the full topology is available. So the current `connect_stream()` would just record the connectivity without actually creating any buffers. The buffers would be created later as part of the 'run_flowgraph()` process, maybe right after `topology::validate()`. 

I was originally thinking that the "hints" would be set in, or returned from, the block `init()` method. Now I'm thinking that it would probably be better to make a new block method, one that makes no assumptions about any runtime state. There could be a default implementation that returns `None`, so existing blocks would not be affected.

**Reply by bastibl** [reply]
> I'm not sure I understand what you are suggesting. Is it to just make it so blocks can fail more gracefully in init() than in work()? And leave it to the user (creator of the flowgraph) to ensure that the buffers are configured properly? Isn't that pretty much the way things are now?

`init()` seemed more natural to me, since (1) it doesn't make sense to start the flowgraph with an invalid configuration and (2) work is  the fast-path and I would avoid checking buffer parameters in each call.

Your idea sounds good. Note that buffer creation is already done during flowgraph startup. The last parameter of `connect_stream_with_type` is a builder. Buffers are only instantiated when the writer is created. (I think that's fine. It should help serializing a flowgraph to instantiate it on another host in the future.)

I think one could make the hints optional parameters of the stream input/output constructors. Maybe we could even have some `_with_constraints` versions of the constructors and allows constraints like min size, exact size, max size, or even buffer types. Not sure if this is really required, though. But having something like `BufferConstraints` or `StreamPortConstraints` could help adding more logic in the future.

**Reply by etschneider** [reply]
Agreed, `work()` is certainly not the place for any of this...

> Note that buffer creation is already done during flowgraph startup. 

That is done in `run_flowgraph()` in  the "connect stream IO" loop, before "init blocks", right?

I don't see any obvious reason that the block `init()` couldn't be done first, so long as there is no expectation that the block's IO is already setup. This would preclude the "failing gracefully" in `init()` option, as the block wouldn't know what the final connection configurations would be at that time.

Regardless, the "hints" need to be communicated from the block instance to the runtime somehow. I see three options:

1. Add "hints" as a block member variable and set it in `block::init()` (As part of `BlockMeta`, `StreamIo`, or a new member)
   - Pro: doesn't change any external interface
   - Con: buffers would not be setup when `init()` is called and would have no way of knowing the final config. 
   - Con (minor): that info would presumably live as long as the block when it is only needed at startup.
2. Add a new `pre_init()`/`get_constraints()` block method
   - Pro: separates concerns. Hints and `init()` are different things at different lifecycle stages.
      -  `init()` can still "fail gracefully" if the final config is not acceptable.
   - Pro (minor): No extra data added to block, when only needed for startup
   - Con (minor?): Changes the block interface, but in a backwards compatible way.
3. Return something from `init()`
   - Con (major): changes the kernel interface in a way that effects all existing blocks.

ATM, I *slightly* prefer option 2 over 1. Maybe I'm missing some better option all together.

If you want to go with option 1, I think a new dedicated `struct BlockConstraints` member would be preferable. Separation of concerns from runtime data (like `StreamIo`), encapsulation of all the "hints" stuff (any future "solver" probably shouldn't need to pick data out of a bunch of random structures), with future extensibility by adding members.

Option 2 would just return the same struct instead of having it a block member.

**Reply by bastibl** [reply]
I think, we could add the `BlockConstraints` to the block (besides sio, mio, meta) and initialize it in the constructor. Otherwise it needs to be wrapped in `Option<BlockConstraints>` and always checked etc. We can just have a `BlockConstraints::default()` that is set with the current `Block` constructor and that means it is unconstrained. And add a `contraints() -> &BlockConstraints` method to the block. Then they can be queried from `topology->blocks` in `run_flowgraph`. I hope, I'm not missing something. Usually, I just give it a try.

**Reply by etschneider** [reply]
I like it!

Do you want me to take a stab at it, or would you rather do it yourself? If it matters, I may be kind of slow, my time is still pretty limited.

**Reply by bastibl** [reply]
Sure, would be great, if you look into it. You can also ping me on Discord, once you start working on it. I'll also put it on my list, but, like you, I don't know when I'll find the time.
